### ISSUE TITLE
Nullptr check failed, when using TensorArray in combination with while_loop and swap_memory

### ISSUE THREAD

**Issue by davzha** [tensorflow/tensorflow#10083]
Environment: TensorFlow-gpu r1.1 build from source on Windows 10

I am using the python API of TensorFlow to train a variant of an LSTM.
For that purpose I use the `tf.while_loop` function to iterate over the time steps
With `swap_memory=True` and not limiting devices to cpu I get the following failure:

`...tensorflow/tensorflow/core/framework/tensor.cc:885] Check failed: nullptr != b.buf_ (nullptr vs. 00...)`

With `swap_memory=False` or limiting devices to cpu this does not happen.

The part of my code, that causes this failure (when commenting it out, it works) is in the body of the while loop:

    ...
    h_gathered = h_ta.gather(tf.range(time))
    h_gathered = tf.transpose(h_gathered, [1, 0, 2])
    syn_t = self.syntactic_weights_ta.read(time)[:, :time]
    syn_t = tf.expand_dims(syn_t, 1)
    syn_state_t = tf.squeeze(tf.tanh(tf.matmul(syn_t, h_gathered)), 1)
    ...

where `time` is zero based and incremented after each step, `h_ta` is a TensorArray

    h_ta = tf.TensorArray(
            dtype=dtype,
            size=max_seq_len,
            clear_after_read=False,
            element_shape=[batch_size, num_hidden],
            tensor_array_name="fw_output")
and `self.syntactic_weights_ta` is also a TensorArray

    self.syntactic_weights_ta = tf.TensorArray(
            dtype=dtype,
            size=max_seq_len,
            tensor_array_name="fw_syntactic_weights")
    self.syntactic_weights_ta = self.syntactic_weights_ta.unstack(syntactic_weights)

What I am trying to achieve in the code snippet is basically a weighted sum over the past outputs, stored in `h_ta`.
In the end I train the network with `tf.train.AdamOptimizer`.

The forward propagation seems to work, as inserting `tf.Print` commands in the sensitive part of the code works.
But I guess the backward propagation 

Unfortunately I could not find out which tensor's buffer points to nullptr.

**Comment by ebrevdo** [comment]
@yuanbyu it's possible that swap_memory takes a tensor that we're keeping track of inside the TensorArray and moves it off the CPU to the GPU.  When the TensorArray tries to access it, it's no longer there.

@davzha we'll need the complete debug log and an example code snippet to reproduce...

**Comment by davzha** [comment]
I was able to reproduce the error running the following script
Since I followed this guide https://github.com/tensorflow/tensorflow/tree/r0.12/tensorflow/contrib/cmake
to build tensorflow-gpu r1.1, it does not support debugging, though if you still require a debug log I'll try to rebuild it.

```
import tensorflow as tf
from tensorflow.contrib.rnn.python.ops.lstm_ops import _lstm_block_cell
import numpy as np


class SALSTMloop(object):
    """Syntax aware LSTM loop, using `_lstm_block_cell`.
    """

    def __init__(self,
                 num_units,
                 input_ta,
                 syntactic_weights_ta,
                 max_sen_len,
                 batch_size,
                 forget_bias=1.0):
        self._num_units = num_units
        self._forget_bias = forget_bias
        self._names = {
            "W": "weights",
            "b": "biases",
            "W_syn": "weight_syntactic",
            "b_syn": "bias_syntactic",
            "wci": "w_i_diag",
            "wco": "w_o_diag",
            "wcf": "w_f_diag",
            "scope": "lstm_cell"
        }
        self.max_sen_len = max_sen_len
        self.zero_output = tf.zeros([batch_size, num_units])
        self.input_ta = input_ta
        self.syntactic_weights_ta = syntactic_weights_ta

    def __call__(self, time, prev_state, h_ta, scope=None):
        with tf.variable_scope(scope or self._names["scope"]):
            x = self.input_ta.read(time)
            input_size = x.get_shape()[1].value

            w = tf.get_variable(self._names["W"], [
                input_size + self._num_units, self._num_units * 4
            ])
            b = tf.get_variable(
                self._names["b"], [w.get_shape().with_rank(2)[1].value],
                initializer=tf.constant_initializer(0.0))
            w_syn = tf.get_variable(self._names["W_syn"], [
                input_size + self._num_units, self._num_units
            ])
            b_syn = tf.get_variable(
                self._names["b_syn"],
                [w_syn.get_shape().with_rank(2)[1].value],
                initializer=tf.constant_initializer(0.0))
            wci = wco = wcf = tf.zeros([self._num_units])

            (cs_prev, h_prev) = prev_state
            (_, cs, _, _, _, _, h) = _lstm_block_cell(
                x,
                cs_prev,
                h_prev,
                w,
                b,
                wci=wci,
                wco=wco,
                wcf=wcf,
                forget_bias=self._forget_bias,
                use_peephole=False)

            h_gathered = h_ta.gather(tf.range(time))
            h_gathered = tf.transpose(h_gathered, [1, 0, 2])
            syn_t = self.syntactic_weights_ta.read(time)[:, :time]

            syn_t = tf.expand_dims(syn_t, 1)

            syn_state_t = tf.squeeze(tf.tanh(tf.matmul(syn_t, h_gathered)), 1)
            gate_syn = tf.matmul(tf.concat([x, h_prev], 1), w_syn) + b_syn
            state_h = tf.multiply(gate_syn, syn_state_t)

            h = h + state_h

            new_state = (cs, h)
            return (time + 1), new_state, h_ta.write(time, h)


max_sen_len = 60
dtype = tf.float32
num_hidden = 512
batch_size = 100
init_state = (
    tf.constant(0.0, dtype=tf.float32, shape=[batch_size, num_hidden]), ) * 2
fw_syn_ta = tf.TensorArray(
    dtype=dtype,
    size=max_sen_len,
    tensor_array_name="syntactic_weights",
    clear_after_read=False)
syntactic_weights = tf.random_uniform([max_sen_len, batch_size, max_sen_len])
fw_syn_ta = fw_syn_ta.unstack(syntactic_weights)
inputs = tf.constant(
    np.arange(batch_size * max_sen_len * max_sen_len),
    shape=[max_sen_len, batch_size, max_sen_len],
    dtype=tf.float32)

fw_input_ta = tf.TensorArray(
    dtype=dtype, size=max_sen_len,
    tensor_array_name="fw_input").unstack(inputs)
fw_output_ta = tf.TensorArray(
    dtype=dtype,
    size=max_sen_len,
    clear_after_read=False,
    element_shape=[batch_size, num_hidden],
    tensor_array_name="fw_output")
fw_loop = SALSTMloop(
    num_units=num_hidden,
    input_ta=fw_input_ta,
    syntactic_weights_ta=fw_syn_ta,
    max_sen_len=max_sen_len,
    batch_size=batch_size)
time = tf.constant(0, dtype=tf.int32, name="time")

_, fw_final_state, fw_final_output_ta = tf.while_loop(
    lambda time, *_: time < max_sen_len,
    fw_loop,
    loop_vars=[time, init_state, fw_output_ta],
    swap_memory=True)
fw_output = fw_final_output_ta.stack()
fw_final_output_ta.close()

final_output = fw_final_output_ta.stack()

loss = tf.reduce_sum(final_output)

train_op = tf.train.AdamOptimizer(learning_rate=0.0001).minimize(loss)

init_op = tf.global_variables_initializer()
with tf.Session() as sess:
    sess.run(init_op)
    output, _ = sess.run([final_output, train_op])
```

**Comment by girving** [comment]
@ebrevdo What's the status of this issue?

**Comment by iansimon** [comment]
I'm seeing the same issue when using tf.map_fn with swap_memory=True.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by thtrieu** [comment]
I am running into this problem, did anyone manage to solve it?

**Comment by tensorflowbutler** [comment]
Nagging Assigneee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 154 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 169 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 184 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 199 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 214 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 229 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 244 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 259 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ebrevdo** [comment]
The error is no longer reproducible with a recent version of TF (e.g. a nightly from the past month) - so I'm assuming the bug has been fixed.
